File: src/episode/api/inventory.py
from __future__ import annotations

import ipaddress
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator, model_validator

from episode.domain.models import CapabilityConfig, Device

RecordingMode = Literal["disabled", "on_event", "on_episode"]
DeviceType = Literal["camera", "doorbell", "alarm_panel", "sensor", "other"]
AuthMode = Literal["digest_wsse", "digest"]


class AreaCreateRequest(BaseModel):
    id: str | None = Field(default=None, max_length=64, pattern=r"^[a-z0-9][a-z0-9_-]*$")
    name: str = Field(min_length=1, max_length=80)
    location: str = Field(default="", max_length=200)

    @field_validator("name", "location")
    @classmethod
    def strip_text(cls, value: str) -> str:
        return value.strip()


class AreaUpdateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    location: str = Field(default="", max_length=200)
    enabled: bool = True

    @field_validator("name", "location")
    @classmethod
    def strip_text(cls, value: str) -> str:
        return value.strip()


class VideoConfigurationRequest(BaseModel):
    enabled: bool = True
    manual_endpoint: bool = False
    protocol: str = Field(default="rtsp", max_length=16)
    port: int | None = Field(default=554, ge=1, le=65535)
    path: str = Field(default="/Streaming/Channels/101", max_length=500)
    recording_mode: RecordingMode = "on_event"


class ONVIFConfigurationRequest(BaseModel):
    enabled: bool = True
    protocol: str = Field(default="http", max_length=16)
    port: int | None = Field(default=80, ge=1, le=65535)
    path: str = Field(default="/onvif/device_service", max_length=500)
    auth_mode: AuthMode = "digest_wsse"
    events_enabled: bool = False
    relaxed_xml: bool = False


class ISAPIConfigurationRequest(BaseModel):
    enabled: bool = False
    protocol: str = Field(default="http", max_length=16)
    port: int | None = Field(default=80, ge=1, le=65535)
    path: str = Field(default="/ISAPI/Event/notification/alertStream", max_length=500)
    ignore_events: list[str] = Field(default_factory=lambda: ["videoloss", "illaccess"])

    @field_validator("ignore_events")
    @classmethod
    def normalize_ignored_events(cls, values: list[str]) -> list[str]:
        return sorted({value.strip().lower() for value in values if value.strip()})


class SDKConfigurationRequest(BaseModel):
    enabled: bool = False
    port: int = Field(default=8000, ge=1, le=65535)


class EpisodePolicyRequest(BaseModel):
    activity_window_seconds: int | None = Field(default=None, ge=1, le=3600)


class DeviceWriteRequest(BaseModel):
    id: str | None = Field(default=None, max_length=64, pattern=r"^[a-z0-9][a-z0-9_-]*$")
    name: str = Field(min_length=1, max_length=80)
    device_type: DeviceType = "camera"
    area_id: str = Field(min_length=1, max_length=64)
    enabled: bool = True
    ip_address: str = Field(default="", max_length=255)
    username: str | None = Field(default=None, max_length=128)
    password: str | None = Field(default=None, max_length=256)
    clear_credentials: bool = False
    episode_policy: EpisodePolicyRequest = Field(default_factory=EpisodePolicyRequest)
    video: VideoConfigurationRequest = Field(default_factory=VideoConfigurationRequest)
    onvif: ONVIFConfigurationRequest = Field(default_factory=ONVIFConfigurationRequest)
    isapi: ISAPIConfigurationRequest = Field(default_factory=ISAPIConfigurationRequest)
    hikvision_sdk: SDKConfigurationRequest = Field(default_factory=SDKConfigurationRequest)

    @field_validator("name", "device_type", "area_id", "ip_address")
    @classmethod
    def strip_text(cls, value: str) -> str:
        return value.strip()

    @model_validator(mode="after")
    def validate_network_configuration(self):
        needs_address = any(
            (
                self.video.enabled,
                self.onvif.enabled,
                self.isapi.enabled,
                self.hikvision_sdk.enabled,
            )
        )
        if needs_address and not self.ip_address:
            raise ValueError("Network address is required for enabled integrations")
        if self.ip_address:
            try:
                ipaddress.ip_address(self.ip_address)
            except ValueError as exc:
                raise ValueError("Network address must be a valid IPv4 or IPv6 address") from exc
        for value in (self.video, self.onvif, self.isapi):
            if value.enabled and value.path and not value.path.startswith("/"):
                raise ValueError("Integration paths must start with '/'")
        if self.video.enabled and not self.onvif.enabled and not self.video.manual_endpoint:
            raise ValueError(
                "Enable a manual RTSP endpoint when video recording is used without ONVIF"
            )
        if self.video.manual_endpoint and (not self.video.protocol or not self.video.path):
            raise ValueError("A manual RTSP endpoint requires a protocol and path")
        return self


SupportStatus = Literal[
    "supported",
    "unsupported",
    "authentication_failed",
    "unreachable",
    "unavailable",
    "not_validated",
]


class IntegrationSupportResponse(BaseModel):
    status: SupportStatus
    summary: str
    checked_at: str | None = None
    capabilities: list[str] = Field(default_factory=list)
    details: dict[str, Any] = Field(default_factory=dict)


class DeviceValidationResponse(BaseModel):
    device_id: str | None = None
    results: dict[str, IntegrationSupportResponse]


class DeviceConfigurationResponse(BaseModel):
    username_configured: bool
    password_configured: bool
    episode_policy: EpisodePolicyRequest
    video: VideoConfigurationRequest
    onvif: ONVIFConfigurationRequest
    isapi: ISAPIConfigurationRequest
    hikvision_sdk: SDKConfigurationRequest


def editable_device_configuration(device: Device) -> dict:
    video = device.get_config("video")
    onvif = device.get_config("onvif")
    isapi = device.get_config("isapi")
    sdk = device.get_config("hikvision_sdk")
    discovered_video = bool(video and video.settings.get("origin") == "onvif")
    manual_video = bool(video and video.protocol and video.path and not discovered_video)
    return DeviceConfigurationResponse(
        username_configured=bool(device.username),
        password_configured=bool(device.password),
        episode_policy=EpisodePolicyRequest(
            activity_window_seconds=device.activity_window_seconds,
        ),
        video=VideoConfigurationRequest(
            enabled=video is not None,
            manual_endpoint=manual_video,
            protocol=video.protocol if manual_video else "rtsp",
            port=video.port if manual_video else 554,
            path=video.path if manual_video else "/Streaming/Channels/101",
            recording_mode=(
                video.settings.get("recording_mode", "on_event") if video else "disabled"
            ),
        ),
        onvif=ONVIFConfigurationRequest(
            enabled=onvif is not None,
            protocol=onvif.protocol if onvif else "http",
            port=onvif.port if onvif else 80,
            path=onvif.path if onvif else "/onvif/device_service",
            auth_mode=onvif.settings.get("auth_mode", "digest_wsse") if onvif else "digest_wsse",
            events_enabled=bool(onvif.settings.get("events_enabled", False)) if onvif else False,
            relaxed_xml=bool(onvif.settings.get("relaxed_xml", False)) if onvif else False,
        ),
        isapi=ISAPIConfigurationRequest(
            enabled=isapi is not None,
            protocol=isapi.protocol if isapi else "http",
            port=isapi.port if isapi else 80,
            path=isapi.path if isapi else "/ISAPI/Event/notification/alertStream",
            ignore_events=list(isapi.settings.get("ignore_events", [])) if isapi else [],
        ),
        hikvision_sdk=SDKConfigurationRequest(
            enabled=sdk is not None,
            port=sdk.port if sdk and sdk.port else 8000,
        ),
    ).model_dump()


def device_from_request(
    device_id: str,
    request: DeviceWriteRequest,
    existing: Device | None = None,
) -> Device:
    managed_capabilities = {"doorbell", "onvif", "isapi", "hikvision_sdk"}
    capabilities = set(existing.capabilities if existing else ()) - managed_capabilities

    configs = {
        key: value
        for key, value in (existing.configs.items() if existing else ())
        if key not in {"video", "onvif", "isapi", "hikvision_sdk"}
    }
    if request.video.enabled:
        previous = existing.get_config("video") if existing else None
        settings = dict(previous.settings) if previous else {}
        settings.pop("origin", None)
        settings.pop("profile_token", None)
        settings["recording_mode"] = request.video.recording_mode
        configs["video"] = CapabilityConfig(
            protocol=request.video.protocol if request.video.manual_endpoint else "",
            port=request.video.port if request.video.manual_endpoint else None,
            path=request.video.path if request.video.manual_endpoint else "",
            settings=settings,
        )
    if request.onvif.enabled:
        previous = existing.get_config("onvif") if existing else None
        settings = dict(previous.settings) if previous else {}
        settings.update(
            auth_mode=request.onvif.auth_mode,
            events_enabled=request.onvif.events_enabled,
            relaxed_xml=request.onvif.relaxed_xml,
        )
        configs["onvif"] = CapabilityConfig(
            protocol=request.onvif.protocol,
            port=request.onvif.port,
            path=request.onvif.path,
            settings=settings,
        )
    if request.isapi.enabled:
        configs["isapi"] = CapabilityConfig(
            protocol=request.isapi.protocol,
            port=request.isapi.port,
            path=request.isapi.path,
            settings={"ignore_events": request.isapi.ignore_events},
        )
    if request.hikvision_sdk.enabled:
        configs["hikvision_sdk"] = CapabilityConfig(port=request.hikvision_sdk.port)

    if request.clear_credentials:
        username = ""
        password = ""
    else:
        username = (
            request.username
            if request.username is not None
            else existing.username
            if existing
            else ""
        )
        password = (
            request.password
            if request.password is not None
            else existing.password
            if existing
            else ""
        )

    return Device(
        id=device_id,
        name=request.name,
        device_type=request.device_type,
        area_id=request.area_id,
        capabilities=sorted(capabilities),
        ip_address=request.ip_address,
        username=username,
        password=password,
        configs=configs,
        activity_window_seconds=request.episode_policy.activity_window_seconds,
        metadata=dict(existing.metadata) if existing else {},
        enabled=request.enabled,
    )


def validation_device_from_request(
    request: DeviceWriteRequest,
    existing: Device | None = None,
) -> Device:
    """Build a probe target while retaining disabled integration endpoint values."""
    device = device_from_request(request.id or "validation", request, existing)
    device.configs["video"] = CapabilityConfig(
        protocol=request.video.protocol if request.video.manual_endpoint else "",
        port=request.video.port if request.video.manual_endpoint else None,
        path=request.video.path if request.video.manual_endpoint else "",
        settings={
            "recording_mode": request.video.recording_mode,
            "manual_endpoint": request.video.manual_endpoint,
        },
    )
    device.configs["onvif"] = CapabilityConfig(
        protocol=request.onvif.protocol,
        port=request.onvif.port,
        path=request.onvif.path,
        settings={
            "auth_mode": request.onvif.auth_mode,
            "events_enabled": request.onvif.events_enabled,
            "relaxed_xml": request.onvif.relaxed_xml,
        },
    )
    device.configs["isapi"] = CapabilityConfig(
        protocol=request.isapi.protocol,
        port=request.isapi.port,
        path=request.isapi.path,
        settings={"ignore_events": request.isapi.ignore_events},
    )
    device.configs["hikvision_sdk"] = CapabilityConfig(port=request.hikvision_sdk.port)
    return device

File: src/episode/ui/inventory.js
import { apiRequest } from "./api.js?v=3";
import { closeDialog, confirmDialog, notify, openDialog } from "./dialogs.js?v=1";
import { escHtml } from "./dom.js";
import { titleCase } from "./format.js";

const selected = value => value ? " selected" : "";
const checked = value => value ? " checked" : "";
const numberOrNull = value => value ? Number(value) : null;

function decodeApiText(value) {
  const element = document.createElement("textarea");
  element.innerHTML = String(value ?? "");
  return element.value;
}

function safeValue(value) {
  return escHtml(decodeApiText(value));
}

function field(formData, name) {
  return String(formData.get(name) || "").trim();
}

function isChecked(formData, name) {
  return formData.get(name) === "on";
}

export function openAreaEditor(area, onSaved) {
  const editing = Boolean(area);
  openDialog({
    title: editing ? "Edit Area" : "Create an Area",
    subtitle: "Areas group related activity and define the current correlation boundary.",
    content: `
      <div class="form-grid">
        <label class="field field-span"><span>Name</span><input name="name" required maxlength="80" value="${safeValue(area?.name || "")}" placeholder="Front entrance"></label>
        <label class="field field-span"><span>Location <small>optional</small></span><input name="location" maxlength="200" value="${safeValue(area?.location || "")}" placeholder="Main gate and approach"></label>
        ${editing ? `<label class="toggle-row field-span"><input type="checkbox" name="enabled"${checked(area.enabled)}><span><strong>Active</strong><small>Disabled Areas remain available to historical Episodes.</small></span></label>` : `
        <details class="form-advanced field-span"><summary>Advanced</summary>
          <label class="field"><span>Area ID <small>generated if empty</small></span><input name="id" maxlength="64" pattern="[a-z0-9][a-z0-9_-]*" placeholder="front-entrance"></label>
        </details>`}
      </div>`,
    submitLabel: editing ? "Save Area" : "Create Area",
    onSubmit: async data => {
      const body = {
        name: field(data, "name"),
        location: field(data, "location"),
        ...(editing ? { enabled: isChecked(data, "enabled") } : { id: field(data, "id") || null }),
      };
      await apiRequest(editing ? `/areas/${encodeURIComponent(area.id)}` : "/areas", {
        method: editing ? "PUT" : "POST",
        body,
      });
      closeDialog();
      notify(editing ? "Area updated" : "Area created");
      await onSaved();
    },
  });
}

function deviceDefaults(device) {
  const config = device?.configuration || {};
  const policy = config.episode_policy || {};
  return {
    episodePolicy: {
      activity_window_seconds: policy.activity_window_seconds ?? 30,
    },
    video: { enabled: true, manual_endpoint: false, protocol: "rtsp", port: 554, path: "/Streaming/Channels/101", recording_mode: "on_event", ...(config.video || {}) },
    onvif: { enabled: true, protocol: "http", port: 80, path: "/onvif/device_service", auth_mode: "digest_wsse", events_enabled: false, relaxed_xml: false, ...(config.onvif || {}) },
    isapi: { enabled: false, protocol: "http", port: 80, path: "/ISAPI/Event/notification/alertStream", ignore_events: ["videoloss", "illaccess"], ...(config.isapi || {}) },
    sdk: { enabled: false, port: 8000, ...(config.hikvision_sdk || {}) },
    reolink: { enabled: false, host: "", port: 9000, media_enabled: false, events_enabled: false, ...(config.reolink || {}) },
  };
}

function integrationToggle(name, title, description, enabled, body, attributes = "") {
  return `<fieldset class="integration-option" data-integration="${name}" ${attributes}>
    <label class="toggle-row integration-toggle">
      <input type="checkbox" name="${name}_enabled"${checked(enabled)}>
      <span><strong>${title}</strong><small>${description}</small></span>
    </label>
    <div class="integration-fields">${body}</div>
  </fieldset>`;
}

const validationIntegrations = [
  ["onvif", "ONVIF"],
  ["isapi", "ISAPI"],
  ["hikvision_sdk", "HCNetSDK"],
  ["reolink", "Reolink"],
];

function renderValidationResults(results = {}) {
  return validationIntegrations.map(([key, label]) => {
    const result = results[key] || {
      status: "not_validated",
      summary: "Support has not been validated",
      capabilities: [],
    };
    const capabilities = (result.capabilities || [])
      .map(capability => `<span>${titleCase(capability)}</span>`).join("");
    return `<div class="validation-result validation-${result.status}">
      <span class="validation-dot"></span>
      <div><strong>${label}</strong><small>${result.summary}</small>
        ${capabilities ? `<div class="validation-capabilities">${capabilities}</div>` : ""}
      </div>
      <span class="validation-status">${titleCase(result.status)}</span>
    </div>`;
  }).join("");
}

function devicePayload(data, editing, device) {
  return {
    id: editing ? device.id : field(data, "id") || null,
    name: field(data, "name"),
    device_type: field(data, "device_type"),
    area_id: field(data, "area_id"),
    enabled: editing ? isChecked(data, "enabled") : true,
    ip_address: field(data, "ip_address"),
    username: field(data, "username") || null,
    password: field(data, "password") || null,
    clear_credentials: isChecked(data, "clear_credentials"),
    episode_policy: {
      activity_window_seconds: Number(field(data, "activity_window_seconds")),
    },
    video: {
      enabled: isChecked(data, "video_enabled"),
      manual_endpoint: isChecked(data, "manual_video_endpoint"),
      protocol: field(data, "video_protocol"),
      port: numberOrNull(field(data, "video_port")),
      path: field(data, "video_path"),
      recording_mode: field(data, "recording_mode"),
    },
    onvif: {
      enabled: isChecked(data, "onvif_enabled"),
      protocol: field(data, "onvif_protocol"),
      port: numberOrNull(field(data, "onvif_port")),
      path: field(data, "onvif_path"),
      auth_mode: field(data, "onvif_auth_mode"),
      events_enabled: isChecked(data, "onvif_events_enabled"),
      relaxed_xml: isChecked(data, "onvif_relaxed_xml"),
    },
    isapi: {
      enabled: isChecked(data, "isapi_enabled"),
      protocol: field(data, "isapi_protocol"),
      port: numberOrNull(field(data, "isapi_port")),
      path: field(data, "isapi_path"),
      ignore_events: field(data, "isapi_ignore_events")
        .split(",").map(value => value.trim()).filter(Boolean),
    },
    hikvision_sdk: {
      enabled: isChecked(data, "hikvision_sdk_enabled"),
      port: Number(field(data, "sdk_port")),
    },
    reolink: {
      enabled: isChecked(data, "reolink_enabled"),
      host: field(data, "reolink_host"),
      port: numberOrNull(field(data, "reolink_port")),
      media_enabled: isChecked(data, "reolink_media_enabled"),
      events_enabled: isChecked(data, "reolink_events_enabled"),
    },
  };
}

export function openDeviceEditor(device, areas, onSaved) {
  const editing = Boolean(device);
  const values = deviceDefaults(device);
  const physicalTypes = ["camera", "doorbell", "alarm_panel", "sensor", "other"];
  const deviceType = physicalTypes.includes(device?.device_type) ? device.device_type : "camera";
  const credentialHint = editing && device.configuration?.password_configured
    ? "Stored securely — leave blank to keep"
    : "Camera password";
  const areaOptions = areas.filter(area => area.enabled || area.id === device?.area_id)
    .map(area => `<option value="${area.id}"${selected(area.id === device?.area_id)}>${area.name}${area.enabled ? "" : " (disabled)"}</option>`).join("");

  const overlay = openDialog({
    title: editing ? "Edit Device" : "Add a Device",
    subtitle: "Start with ONVIF. Vendor integrations can complement it when useful.",
    wide: true,
    content: `
      <div class="form-section"><h3>Device</h3><div class="form-grid">
        <label class="field"><span>Name</span><input name="name" required maxlength="80" value="${safeValue(device?.name || "")}" placeholder="Front door camera"></label>
        <label class="field"><span>Area</span><select name="area_id" required><option value="">Choose an Area</option>${areaOptions}</select></label>
        <label class="field"><span>Network address</span><input name="ip_address" inputmode="decimal" value="${safeValue(device?.ip_address || "")}" placeholder="Optional for externally driven sensors"><small>Required only when Episode connects directly to this Device.</small></label>
        <label class="field"><span>Device type</span><select name="device_type" data-device-type>
          ${physicalTypes.map(type => `<option value="${type}"${selected(deviceType === type)}>${titleCase(type)}</option>`).join("")}
        </select><small>Physical role only. Manufacturer and vendor integrations are kept separate.</small></label>
        ${editing ? `<label class="toggle-row field-span"><input type="checkbox" name="enabled"${checked(device.enabled)}><span><strong>Active</strong><small>Disable without losing historical relationships.</small></span></label>` : ""}
      </div></div>

      <div class="form-section"><h3>Credentials</h3><div class="form-grid">
        <label class="field"><span>Username</span><input name="username" autocomplete="off" placeholder="${editing && device.configuration?.username_configured ? "Stored — leave blank to keep" : "admin"}"></label>
        <label class="field"><span>Password</span><input type="password" name="password" autocomplete="new-password" placeholder="${credentialHint}"></label>
        ${editing && (device.configuration?.username_configured || device.configuration?.password_configured) ? `<label class="toggle-row field-span"><input type="checkbox" name="clear_credentials"><span><strong>Clear stored credentials</strong><small>Use only for devices that allow anonymous access.</small></span></label>` : ""}
      </div></div>

      <div class="form-section"><h3>Capture</h3>
        <div class="form-grid capture-policy-fields">
          <label class="field"><span>Episode activity window</span><input name="activity_window_seconds" type="number" min="1" max="3600" required value="${values.episodePolicy.activity_window_seconds}"><small>Seconds this Device keeps an Episode open after each Event. Other recording Devices follow the Episode.</small></label>
        </div>
        ${integrationToggle("video", "Video recording", "Capture this Device when its Area is active.", values.video.enabled, `
          <div class="form-grid">
            <label class="field"><span>Recording behavior</span><select name="recording_mode">
              <option value="disabled"${selected(values.video.recording_mode === "disabled")}>Do not record</option>
              <option value="on_event"${selected(values.video.recording_mode === "on_event")}>Own Events only</option>
              <option value="on_episode"${selected(values.video.recording_mode === "on_episode")}>Any Episode in this Area</option>
            </select></label>
          </div>`) }
      </div>

      <div class="form-section">
        <div class="form-section-heading"><h3>Integrations</h3><p>Configured connections activate automatically when the Device is saved.</p></div>
        <div class="integration-stack">
          ${integrationToggle("onvif", "ONVIF", "Standards-based discovery, media, and optional Events.", values.onvif.enabled, `
            <label class="toggle-row"><input type="checkbox" name="onvif_events_enabled"${checked(values.onvif.events_enabled)}><span><strong>Receive ONVIF Events</strong><small>Disabled by default to avoid noisy motion state changes.</small></span></label>
            <label class="toggle-row"><input type="checkbox" name="onvif_relaxed_xml"${checked(values.onvif.relaxed_xml)}><span><strong>Tolerate malformed SOAP XML</strong><small>Compatibility fallback for Devices that return malformed ONVIF responses. Leave off unless validation fails.</small></span></label>`) }
          <div class="integration-group-label"><strong>Hikvision enhancements</strong><span>Optional vendor connections that complement ONVIF.</span></div>
          ${integrationToggle("isapi", "ISAPI Event stream", "Rich motion and classification Events. It is currently active when this switch is on.", values.isapi.enabled, "")}
          ${integrationToggle("hikvision_sdk", "HCNetSDK", "Native callbacks for doorbell rings and door-control Events.", values.sdk.enabled, `
            <div class="role-guidance">Available for Doorbell Devices.</div>`) }
          <div class="integration-group-label"><strong>Reolink</strong><span>Native binary protocol for Reolink cameras.</span></div>
          ${integrationToggle("reolink", "Reolink API", "Discovery, media, and Events over the Reolink binary protocol.", values.reolink.enabled, `
            <label class="toggle-row"><input type="checkbox" name="reolink_media_enabled"${checked(values.reolink.media_enabled)}><span><strong>Enable media (streams &amp; snapshots)</strong><small>Register the discovered RTSP stream and binary snapshots so recording and snapshot-on-event work without ONVIF.</small></span></label>
            <label class="toggle-row"><input type="checkbox" name="reolink_events_enabled"${checked(values.reolink.events_enabled)}><span><strong>Receive Reolink events</strong><small>Listen for motion and detection events pushed over the binary protocol. Disabled by default to avoid noisy state changes.</small></span></label>`) }
        </div>
        <div class="validation-panel">
          <div class="validation-heading">
            <div><strong>Device validation</strong><span>Checks support without enabling integrations.</span></div>
            <button type="button" class="button button-ghost" data-validate-device>Validate and discover</button>
          </div>
          <div class="validation-results" data-validation-results>
            ${renderValidationResults(device?.integration_support)}
          </div>
        </div>
      </div>

      <details class="form-advanced"><summary>Manual connection overrides</summary>
        <p class="configuration-note">These values are configured endpoints or fallbacks. Runtime-discovered manufacturer, model, firmware, media profiles, and selected profile are read-only on the Device page.</p>
        <div class="advanced-grid">
          <fieldset data-manual-video>
            <legend>RTSP fallback</legend>
            <label class="toggle-row"><input type="checkbox" name="manual_video_endpoint"${checked(values.video.manual_endpoint)}><span><strong>Use manual endpoint</strong><small>ONVIF-discovered media is preferred when available.</small></span></label>
            <div class="manual-endpoint-fields">
              <label class="field"><span>Protocol</span><input name="video_protocol" value="${safeValue(values.video.protocol)}"></label>
              <label class="field"><span>Port</span><input name="video_port" type="number" min="1" max="65535" value="${values.video.port || ""}"></label>
              <label class="field"><span>Path</span><input name="video_path" value="${safeValue(values.video.path)}"></label>
            </div>
          </fieldset>
          <fieldset><legend>ONVIF service</legend><label class="field"><span>Protocol</span><input name="onvif_protocol" value="${safeValue(values.onvif.protocol)}"></label><label class="field"><span>Port</span><input name="onvif_port" type="number" min="1" max="65535" value="${values.onvif.port || ""}"></label><label class="field"><span>Path</span><input name="onvif_path" value="${safeValue(values.onvif.path)}"></label><label class="field"><span>Authentication</span><select name="onvif_auth_mode"><option value="digest_wsse"${selected(values.onvif.auth_mode === "digest_wsse")}>Digest + WS-Username Token</option><option value="digest"${selected(values.onvif.auth_mode === "digest")}>Digest only</option></select></label></fieldset>
          <fieldset><legend>ISAPI endpoint</legend><label class="field"><span>Protocol</span><input name="isapi_protocol" value="${safeValue(values.isapi.protocol)}"></label><label class="field"><span>Port</span><input name="isapi_port" type="number" min="1" max="65535" value="${values.isapi.port || ""}"></label><label class="field"><span>Path</span><input name="isapi_path" value="${safeValue(values.isapi.path)}"></label><label class="field"><span>Ignored Events</span><input name="isapi_ignore_events" value="${safeValue((values.isapi.ignore_events || []).join(", "))}"></label></fieldset>
          <fieldset><legend>HCNetSDK login</legend><label class="field"><span>SDK port</span><input name="sdk_port" type="number" min="1" max="65535" value="${values.sdk.port}"></label></fieldset>
          <fieldset><legend>Reolink API</legend><label class="field"><span>API host</span><input name="reolink_host" value="${safeValue(values.reolink.host)}" placeholder="Defaults to the Device address"><small>Optional. Overrides the Device network address.</small></label><label class="field"><span>API port</span><input name="reolink_port" type="number" min="1" max="65535" value="${values.reolink.port || ""}"></label></fieldset>
        </div>
        ${editing ? "" : `<label class="field"><span>Device ID <small>generated if empty</small></span><input name="id" maxlength="64" pattern="[a-z0-9][a-z0-9_-]*" placeholder="front-door-camera"></label>`}
      </details>`,
    submitLabel: editing ? "Save Device" : "Add Device",
    onSubmit: async data => {
      const payload = devicePayload(data, editing, device);
      await apiRequest(editing ? `/devices/${encodeURIComponent(device.id)}` : "/devices", {
        method: editing ? "PUT" : "POST", body: payload,
      });
      closeDialog();
      notify(editing ? "Device and integrations updated" : "Device added and integrations activated");
      await onSaved();
    },
  });

  let validationResults = { ...(device?.integration_support || {}) };
  const updateIntegration = option => {
    const toggle = option.querySelector(".integration-toggle input");
    option.classList.toggle("integration-disabled", !toggle.checked);
  };
  overlay.querySelectorAll("[data-integration]").forEach(option => {
    const toggle = option.querySelector(".integration-toggle input");
    toggle.addEventListener("change", () => updateIntegration(option));
    updateIntegration(option);
  });

  const typeSelect = overlay.querySelector("[data-device-type]");
  const sdkOption = overlay.querySelector('[data-integration="hikvision_sdk"]');
  const sdkToggle = sdkOption.querySelector(".integration-toggle input");
  const sdkWasConfigured = values.sdk.enabled;
  const updateDeviceRole = () => {
    if (!editing && !["camera", "doorbell"].includes(typeSelect.value)) {
      for (const integration of ["video", "onvif", "isapi", "hikvision_sdk", "reolink"]) {
        const option = overlay.querySelector(`[data-integration="${integration}"]`);
        const toggle = option.querySelector(".integration-toggle input");
        toggle.checked = false;
        updateIntegration(option);
      }
    }
    const roleAvailable = typeSelect.value === "doorbell" || sdkWasConfigured;
    const supported = validationResults.hikvision_sdk?.status !== "unsupported";
    const available = roleAvailable && supported;
    sdkToggle.disabled = !available;
    if (!available) sdkToggle.checked = false;
    sdkOption.classList.toggle("integration-role-unavailable", !available);
    updateIntegration(sdkOption);
  };

  const resultsElement = overlay.querySelector("[data-validation-results]");
  const applyValidation = () => {
    resultsElement.innerHTML = renderValidationResults(validationResults);
    for (const integration of ["onvif", "isapi", "reolink"]) {
      const option = overlay.querySelector(`[data-integration="${integration}"]`);
      const toggle = option.querySelector(".integration-toggle input");
      const unsupported = validationResults[integration]?.status === "unsupported";
      toggle.disabled = unsupported;
      if (unsupported) toggle.checked = false;
      option.classList.toggle("integration-support-unsupported", unsupported);
      updateIntegration(option);
    }
    updateDeviceRole();
  };
  typeSelect.addEventListener("change", updateDeviceRole);
  applyValidation();

  const form = overlay.querySelector("form");
  const validateButton = overlay.querySelector("[data-validate-device]");
  validateButton.addEventListener("click", async () => {
    if (!form.reportValidity()) return;
    const originalLabel = validateButton.textContent;
    validateButton.disabled = true;
    validateButton.textContent = "Validating…";
    try {
      const response = await apiRequest("/devices/validate", {
        method: "POST",
        body: devicePayload(new FormData(form), editing, device),
      });
      validationResults = response.results;
      applyValidation();
      notify("Device validation completed");
    } catch (error) {
      notify(`Validation failed: ${error.message}`, "warning");
    } finally {
      validateButton.disabled = false;
      validateButton.textContent = originalLabel;
    }
  });

  const manualVideo = overlay.querySelector("[data-manual-video]");
  const manualVideoToggle = manualVideo.querySelector('[name="manual_video_endpoint"]');
  const updateManualVideo = () => manualVideo.classList.toggle("manual-endpoint-disabled", !manualVideoToggle.checked);
  manualVideoToggle.addEventListener("change", updateManualVideo);
  updateManualVideo();
}

export function confirmAreaDelete(area, onDeleted) {
  confirmDialog({
    title: `Delete ${decodeApiText(area.name)}?`,
    message: "Only unused Areas can be deleted. Areas with Devices or incident history must be disabled.",
    onConfirm: async () => {
      await apiRequest(`/areas/${encodeURIComponent(area.id)}`, { method: "DELETE" });
      closeDialog(); notify("Area deleted"); await onDeleted();
    },
  });
}

export function confirmDeviceDelete(device, onDeleted) {
  confirmDialog({
    title: `Delete ${decodeApiText(device.name)}?`,
    message: "Only Devices without incident history can be deleted. Otherwise disable the Device to preserve its relationships.",
    onConfirm: async () => {
      await apiRequest(`/devices/${encodeURIComponent(device.id)}`, { method: "DELETE" });
      closeDialog(); notify("Device deleted and integrations updated"); await onDeleted();
    },
  });
}
